Rework the Greentea docs (#927)
ARMmbed/Handbook: docs/tools/testing/testing_greentea.md

## Greentea testing applications

> Greentea is the automated testing tool for Arm Mbed OS development. It's a test runner that automates the process of flashing development boards, starting tests, and accumulating test results into test reports. Developers use it for local development as well as for automation in a Continuous Integration environment.

Reviewer: Is there some reason to use capital first letters for "Continous Integration"?

> 

Reviewer 2: t

If you need to test with multiple configurations, then you can use the `--app-config` option. This overrides the search for an `mbed_app.json` file and uses the configuration file that you specify for the build.

### Writing your first test

You can write tests for your own project or add more tests to Mbed OS. You can write tests by using the [Greentea client](https://github.com/ARMmbed/mbed-os/tree/master/features/frameworks/greentea-client), and the [UNITY](https://github.com/ARMmbed/mbed-os/tree/master/features/frameworks/unity) and [utest](https://github.com/ARMmbed/mbed-os/tree/master/features/frameworks/utest) frameworks, which are located in `/features/frameworks`.

Let's write your first test. Use Mbed CLI to create a new project:

```
$ mbed new first-greentea-test
```

> As specified above, there is a convention where all tests live in the `TESTS/` directory. In the `first-greentea-test` folder create a folder `TESTS/tests/simple-test/`.

Reviewer 3: I'd suggest the following example folder structure since the example kind of looks like a duplication (double "tests" folders):
```suggestion
As specified above, there is a convention where all tests live in the `TESTS/` directory. In the `first-greentea-test` folder create a folder `TESTS/test-group/simple-test/`.
```

> ![Exporting to Make](https://s3-us-west-2.amazonaws.com/mbed-os-docs-images/test01.png)

Reviewer 4: We should promote working with Mbed CLI within it's own documentation where possible. Either this is titled incorrectly, or it should be removed (I can't see the image content)

Author: Hmm, I think I copied something wrong there. The description just shows the tree structure, nothing to do with make. My bad :-)

> ![Tree structure for Greentea tests](https://s3-us-west-2.amazonaws.com/mbed-os-docs-images/test01.png)

Reviewer 3: Do we usually host these on s3? Perhaps just using an text representation of the structure would be easier. Ex:

```
project/
    TESTS/
        test-group/
            simple-test/
                main.cpp

Reviewer: AccessDenied: https://s3-us-west-2.amazonaws.com/mbed-os-docs-images/test01.png

Author: Yeah, the image is in this pr, but not uploaded to s3 yet. @bridadan Good suggestion, I'll update it to use that.

In this folder, create a file `main.cpp`. In here you can use UNITY, utest and the Greentea Client to write your test:

```cpp
#include "mbed.h"
#include "utest/utest.h"
#include "unity/unity.h"
#include "greentea-client/test_env.h"

using namespace utest::v1;

> // this is how a test looks

Reviewer 3: Nit: differentiate between "test" with "test case"
```suggestion
// this is how a test case looks
```

static control_t simple_test(const size_t call_count) {
    /* test content here */
    TEST_ASSERT_EQUAL(4, 2 * 2);

    return CaseNext;
}

utest::v1::status_t greentea_setup(const size_t number_of_cases) {

>     // here we specify the timeout (60s) and the host runner (the name of our Python file)

Reviewer 3: Nit: `default_auto` is currently provided by `htrun`, not by a python file within mbed-os.

```suggestion
    // here we specify the timeout (60s) and the host test (a built-in host test or the name of our Python file)
```

>     GREENTEA_SETUP(1*60, "default_auto");

Reviewer 3: Nit: Ok to drop the multiplication here?
```suggestion
    GREENTEA_SETUP(60, "default_auto");
```

    return greentea_test_setup_handler(number_of_cases);
}

> // list of all cases

Reviewer: This is listing test cases which are contained in this file and there are multiple files. "all" sounds me a bit more than one file tests

#### Creating the Greentea test

Now let's write the embedded part of this test. Create a new file `main.cpp` in `TESTS/tests/integration-test`, and fill it with:

```cpp
#include "mbed.h"
#include "utest/utest.h"
#include "unity/unity.h"
#include "greentea-client/test_env.h"

using namespace utest::v1;

static control_t hello_world_test(const size_t call_count) {
    // send a message to the host runner
    greentea_send_kv("init", "hello");

    // wait until we get a message back
    // if this takes too long the timeout will trigger, so no need to handle this here
    char _key[20], _value[128];
    while (1) {
        greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));

        // check if the key equals init, and if the return value is 'world'
        if (strcmp(_key, "init") == 0) {
            TEST_ASSERT_EQUAL(0, strcmp(_value, "world"));
            break;
        }
    }

   return CaseNext;
}

utest::v1::status_t greentea_setup(const size_t number_of_cases) {
   // here we specify the timeout (60s) and the host runner (the name of our Python file)

>    GREENTEA_SETUP(1*60, "hello_world_tests");

Reviewer 3: Same nit pick about the multiplication here.
```suggestion
   GREENTEA_SETUP(60, "hello_world_tests");
```

Reviewer 3: One other `1*60` that can be removed

```cpp
#include "mbed.h"
#include "utest/utest.h"
#include "unity/unity.h"
#include "greentea-client/test_env.h"

using namespace utest::v1;

// This is how a test case looks
static control_t simple_test(const size_t call_count) {
    /* test content here */
    TEST_ASSERT_EQUAL(4, 2 * 2);

    return CaseNext;
}

utest::v1::status_t greentea_setup(const size_t number_of_cases) {
    // Here, we specify the timeout (60s) and the host test (a built-in host test or the name of our Python file)
    GREENTEA_SETUP(60, "default_auto");

    return greentea_test_setup_handler(number_of_cases);
}

// List of test cases in this file
Case cases[] = {
    Case("simple test", simple_test)
};

Specification specification(greentea_setup, cases);

int main() {
    return !Harness::run(specification);
}
```

#### Running the test

<span class="tips">**Tip:** To see all tests, run `mbed test --compile-list`.</span>

Run the test:

```
# run the test with the GCC_ARM toolchain, automatically detect the target, and run in verbose mode (-v)
$ mbed test -t GCC_ARM -m auto -v -n tests-test-group-simple-test
```

> This yields (on a K64F):

Reviewer 2: Can we please use a different board for our examples?

#### Limiting tests

You can select test cases by name using the `-n` argument. This command executes all tests named `tests-mbedmicro-rtos-mbed-mail` from all builds in the test specification:

```
$ mbed test -n tests-mbedmicro-rtos-mbed-mail
```

When using the `-n` argument, you can use the `*` character at the end of a test name to match all tests that share a prefix. This command executes all tests that start with `tests-mbedmicro-rtos`:

```
$ mbed test -n tests-mbedmicro-rtos-*
```

You can use a comma (`,`) to separate test names (argument `-n`) and build names (argument `-t`). This command executes the tests `tests-mbedmicro-rtos-mbed-mail` and `tests-mbed_drivers-c_strings` for the `K64F-ARM` and `K64F-GCC_ARM` builds in the test specification:

```
$ mbed test -n tests-mbedmicro-rtos-mbed-mail,tests-mbed_drivers-c_strings -t K64F-ARM,K64F-GCC_ARM
```

#### Selecting platforms

You can limit which boards Greentea uses for testing by using the `--use-tids` argument.

```
$ mbed test --use-tids 02400203C3423E603EBEC3D8,024002031E031E6AE3FFE3D2
```

Where `02400203C3423E603EBEC3D8` and `024002031E031E6AE3FFE3D` are the target IDs of platforms attached to your system.

> You can view target IDs using the [`mbed-ls`](https://github.com/ARMmbed/mbed-os-tools/tree/master/packages/mbed-ls) tool (installed with Greentea).

Reviewer 4: Please use `mbed detect` instead.

Author: Updated. Copied it from old content but missed this.

For an explanation of the arguments used in this command, please run `mbedhtrun --help`.

### Command-line usage

This section highlights a few of the capabilities of the Greentea command-line interface. For a full list of the available options, please run `mbed test --help`.

#### Listing all tests

You can use the `-l` argument to list all available tests:

```

> $ mbed test -l

Reviewer 3: This will technically work, but it isn't really the supported interface (I can go into more details if you like, but its a bit messy).

If you want the list of compilable tests, you want `mbed test --compile-list`. (https://github.com/ARMmbed/mbed-cli/blob/master/mbed/mbed.py#L2744)
If you want the list of tests that are compiled and ready to run, you wan `mbed test --run-list` (https://github.com/ARMmbed/mbed-cli/blob/master/mbed/mbed.py#L2746)

[mbed] Working path "/Users/janjon01/repos/first-greentea-test" (program)
Test Case:
    Name: mbed-os-components-storage-blockdevice-component_flashiap-tests-filesystem-fopen
    Path: ./mbed-os/components/storage/blockdevice/COMPONENT_FLASHIAP/TESTS/filesystem/fopen
Test Case:
    Name: mbed-os-features-cellular-tests-api-cellular_device
    Path: ./mbed-os/features/cellular/TESTS/api/cellular_device

...
```

#### Executing all tests

The default action of Greentea using `mbed test` is to execute all tests that are found. You can also add `-v` to make the output more verbose.

#### Limiting tests

You can select test cases by name using the `-n` argument. This command executes all tests named `tests-mbedmicro-rtos-mbed-mail` from all builds in the test specification:

```
$ mbed test -n tests-mbedmicro-rtos-mbed-mail
```

> When using the `-n` argument, you can use the `*` character at the end of a test name to match all tests that share a prefix. This command executes all tests that start with `tests-mbedmicro-rtos`:

Reviewer 3: This will now use "globbing" rules, so you can actually sprinkle the `*` characters any where you like.

```
$ mbed test -n tests-mbedmicro-rtos-*
```

You can use a comma (`,`) to separate test names (argument `-n`) and build names (argument `-t`). This command executes the tests `tests-mbedmicro-rtos-mbed-mail` and `tests-mbed_drivers-c_strings` for the `K64F-ARM` and `K64F-GCC_ARM` builds in the test specification:

```
$ mbed test -n tests-mbedmicro-rtos-mbed-mail,tests-mbed_drivers-c_strings -t K64F-ARM,K64F-GCC_ARM
```

#### Selecting platforms

You can limit which boards Greentea uses for testing by using the `--use-tids` argument.

```

> $ mbed test --use-tids 02400203C3423E603EBEC3D8,024002031E031E6AE3FFE3D2

Reviewer 3: This will currently error unless you add `--run`.

```

Where `02400203C3423E603EBEC3D8` and `024002031E031E6AE3FFE3D` are the target IDs of platforms attached to your system.

You can view target IDs using Mbed CLI:

```
$ mbed detect
+--------------+---------------------+------------+------------+-------------------------+
|platform_name |platform_name_unique |mount_point |serial_port |target_id                |
+--------------+---------------------+------------+------------+-------------------------+
|K64F          |K64F[0]              |E:          |COM160      |024002031E031E6AE3FFE3D2 |
|K64F          |K64F[1]              |F:          |COM162      |02400203C3423E603EBEC3D8 |
|LPC1768       |LPC1768[0]           |G:          |COM5        |1010ac87cfc4f23c4c57438d |

> +--------------+---------------------+------------+------------+-------------------------+

Reviewer 3: If your current working directly is within an Mbed OS project, the output of `mbed detect` looks like this:

```
mbed detect
[mbed] Working path "C:\Users\bridan01\onedrive_arm\Documents\dev\mbed-os" (program)

[mbed] Detected K64F, port COM79, mounted F:, interface version 0250:
[mbed] Supported toolchains for K64F
| Target        | mbed OS 2 | mbed OS 5 | uARM |    IAR    |    ARM    |  GCC_ARM  |
|---------------|-----------|-----------|------|-----------|-----------|-----------|
| K64F          | Supported | Supported |  -   | Supported | Supported | Supported |
| RAPIDIOT_K64F | Supported | Supported |  -   | Supported | Supported | Supported |
Supported targets: 2

[mbed] Detected K64F, port COM78, mounted D:, interface version 0251:
[mbed] Supported toolchains for K64F
| Target        | mbed OS 2 | mbed OS 5 | uARM |    IAR    |    ARM    |  GCC_ARM  |
|---------------|-----------|-----------|------|-----------|-----------|-----------|
| K64F          | Supported | Supported |  -   | Supported | Supported | Supported |
| RAPIDIOT_K64F | Supported | Supported |  -   | Supported | Supported | Supported |
Supported targets: 2
```
Which unfortunately does not list the target id. `mbedls` will but I know we want to direct folks to use Mbed CLI commands. In this case I don't think we have many options here, any ideas @theotherjimmy?

Reviewer 4: Oh, huh. Not only does `mbed detect` omit the information, it also neglects to use unique names. I suppose we'll have to use `mbedls` until we fix this in `mbed detect`.

```

In this case, you won't test one target, the LPC1768.

#### Creating reports

Greentea supports a number of report formats.

##### HTML

This creates an interactive HTML page with test results and logs.

```
mbed test --report-html html_report.html
```

##### JUnit

This creates an XML JUnit report, which you can use with popular Continuous Integration software, such as [Jenkins](https://jenkins.io/index.html).

```
mbed test --report-junit junit_report.xml
```

##### JSON

This creates a general JSON report.

```
mbed test --report-json json_report.json
```

##### Plain text

This creates a human-friendly text summary of the test run.

```

> mbed test --report-text text_report.text

Reviewer 3: All of these report commands would need the `--run` command added first, otherwise it would throw an exception:

```
$ mbed test -n tests-mbed_functional-callback -v --report-json report.json
[mbed] Working path "C:\Users\bridan01\onedrive_arm\Documents\dev\mbed-os" (program)
[mbed] Exec "c:\python27\python.exe -u C:\Users\bridan01\onedrive_arm\Documents\dev\mbed-os\tools\test.py -t GCC_ARM -m K64F --source . --build .\BUILD\tests\K64F\GCC_ARM --test-spec .\BUILD\tests\K64F\GCC_ARM\test_spec.json --build-data .\BUILD\tests\K64F\GCC_ARM\build_data.json -n tests-mbed_functional-callback -v --greentea --report-json report.json" in "C:\Users\bridan01\onedrive_arm\Documents\dev\mbed-os"
usage: test.py [-h] [-m MCU] [--custom-targets CUSTOM_TARGETS_DIRECTORY]
               [-t TOOLCHAIN] [--color] [--cflags CFLAGS]
               [--asmflags ASMFLAGS] [--ldflags LDFLAGS] [-c]
               [--profile PROFILE] [--app-config APP_CONFIG] [-D MACROS]
               [-j JOBS] [--source SOURCE_DIR] [--build BUILD_DIR] [-l]
               [-p PATHS] [-f FORMAT] [--continue-on-build-fail] [-n NAMES]
               [--test-config TEST_CONFIG] [--test-spec TEST_SPEC]
               [--build-report-junit BUILD_REPORT_JUNIT]
               [--build-data BUILD_DATA] [-v] [--stats-depth STATS_DEPTH]
               [--ignore IGNORE] [--icetea] [--greentea]
test.py: error: unrecognized arguments: --report-json report.json
[mbed] ERROR: "c:\python27\python.exe" returned error.
       Code: 2
       Path: "C:\Users\bridan01\onedrive_arm\Documents\dev\mbed-os"
       Command: "c:\python27\python.exe -u C:\Users\bridan01\onedrive_arm\Documents\dev\mbed-os\tools\test.py -t GCC_ARM -m K64F --source . --build .\BUILD\tests\K64F\GCC_ARM --test-spec .\BUILD\tests\K64F\GCC_ARM\test_spec.json --build-data .\BUILD\tests\K64F\GCC_ARM\build_data.json -n tests-mbed_functional-callback -v --greentea --report-json report.json"
       Tip: You could retry the last command with "-v" flag for verbose output
---
```

                    "binaries": [
                        {
                            "binary_type": "bootable",
                            "path": "./BUILD/K64F/ARM/tests-mbedmicro-rtos-mbed-mail.bin"
                        }
                    ]
                },
                "tests-mbed_drivers-c_strings": {
                    "binaries": [
                        {
                            "binary_type": "bootable",
                            "path": "./BUILD/K64F/ARM/tests-mbed_drivers-c_strings.bin"
                        }
                    ]
                }
            }
        },
        "K64F-GCC": {
            "platform": "K64F",
            "toolchain": "GCC_ARM",
            "base_path": "./BUILD/K64F/GCC_ARM",
            "baud_rate": 9600,
            "tests": {
                "tests-mbedmicro-rtos-mbed-mail": {
                    "binaries": [
                        {
                            "binary_type": "bootable",
                            "path": "./BUILD/K64F/GCC_ARM/tests-mbedmicro-rtos-mbed-mail.bin"
                        }
                    ]
                }
            }
        }
    }
}
```

> If you now run `mbed test -l` this will now list only these tests:

Reviewer 3: You want to use the `--run-list` option here as well.

```suggestion
If you now run `mbed test --run-list` this will now list only these tests:
```

#### Executing all tests

The default action of Greentea using `mbed test` is to execute all tests that are found. You can also add `-v` to make the output more verbose.

#### Limiting tests

You can select test cases by name using the `-n` argument. This command executes all tests named `tests-mbedmicro-rtos-mbed-mail` from all builds in the test specification:

```
$ mbed test -n tests-mbedmicro-rtos-mbed-mail
```

When using the `-n` argument, you can use the `*` character as a wildcard. This command executes all tests that start with `tests-` and have `-rtos-` in them.

```
$ mbed test -n tests-*-rtos-*
```

You can use a comma (`,`) to separate test names (argument `-n`) and build names (argument `-t`). This command executes the tests `tests-mbedmicro-rtos-mbed-mail` and `tests-mbed_drivers-c_strings` for the `K64F-ARM` and `K64F-GCC_ARM` builds in the test specification:

```
$ mbed test -n tests-mbedmicro-rtos-mbed-mail,tests-mbed_drivers-c_strings -t K64F-ARM,K64F-GCC_ARM
```

#### Selecting platforms

You can limit which boards Greentea uses for testing by using the `--use-tids` argument.

```

> $ mbed test --use-tids 02400203C3423E603EBEC3D8,024002031E031E6AE3FFE3D2 --run

Reviewer 3: Same thing that I mentioned below about the reporting section, this isn't in the `mbed test --help` text. How do you feel about removing this?

Author: Well, isn't it great that our docs are more complete than our CLI?

+--------------+---------------------+------------+------------+-------------------------+
|K64F          |K64F[0]              |E:          |COM160      |024002031E031E6AE3FFE3D2 |
|K64F          |K64F[1]              |F:          |COM162      |02400203C3423E603EBEC3D8 |
|LPC1768       |LPC1768[0]           |G:          |COM5        |1010ac87cfc4f23c4c57438d |
+--------------+---------------------+------------+------------+-------------------------+
```

In this case, you won't test one target, the LPC1768.

#### Creating reports

Greentea supports a number of report formats.

##### HTML

This creates an interactive HTML page with test results and logs.

```
mbed test --report-html html_report.html --run
```

##### JUnit

This creates an XML JUnit report, which you can use with popular Continuous Integration software, such as [Jenkins](https://jenkins.io/index.html).

```
mbed test --report-junit junit_report.xml --run
```

##### JSON

This creates a general JSON report.

```
mbed test --report-json json_report.json --run
```

##### Plain text

This creates a human-friendly text summary of the test run.

```

> mbed test --report-text text_report.text --run

Reviewer 3: Thinking about this more, none of the report arguments are currently part of the help options in Mbed CLI. This only works because Mbed CLI passes along arguments to the underlying scripts when it doesn't recognize it. How do you feel about removing this reporting section altogether?
